logstash-core/src/main/java/org/logstash/plugins/internal/Common.java

```diff
@@ -0,0 +1,126 @@
+package org.logstash.plugins.internal;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class Common {
+    public static class AddressState {
+        private final String address;
+        private final Set<InternalOutput> outputs = ConcurrentHashMap.newKeySet();
+        private volatile InternalInput input = null;
+
+        private AddressState(String address) {
+            this.address = address;
+            this.input = input;
+        }
+
+        public Set<InternalOutput> getOutputs() {
+            return outputs;
+        }
+
+        public InternalInput getInput() {
+            return input;
+        }
+    }
+
+    public static ConcurrentHashMap<String, AddressState> ADDRESS_STATES = new ConcurrentHashMap<>();
+
+    static class AddressesByRunState {
+        private final List<String> running;
+        private final List<String> notRunning;
+
+        AddressesByRunState() {
+            this.running = new ArrayList<>();
+            this.notRunning = new ArrayList<>();
+        }
+
+        public List<String> getRunning() {
+            return running;
+        }
+
+        public List<String> getNotRunning() {
+            return notRunning;
+        }
+    }
+
+    static AddressesByRunState addressesByRunState() {
+        AddressesByRunState addressesByRunState = new AddressesByRunState();
+        ADDRESS_STATES.forEach( (address, state) -> {
+            if (state.input != null && state.input.isRunning()) {
+                addressesByRunState.running.add(address);
+            } else {
+                addressesByRunState.notRunning.add(address);
+            }
+        });
+        return addressesByRunState;
+    }
+
+    public static void registerSender(Collection<String> addresses, InternalOutput output) {
```

Comment:
nitpick: this would read clearer to me if the arguments were reversed; they can also be final:

~~~ diff
-    public static void registerSender(Collection<String> addresses, InternalOutput output) {
+    public static void registerSender(final InternalOutput output, final Collection<String> addresses) {
~~~

Reply:
I agree! I wound up at this weird order guessing at why JRuby couldn't find these methods. Good catch.